grease/src/Grease/Syntax.hs

```diff
@@ -56,3 +83,112 @@ parsedProgramCfgMap prog =
       cfgWithName c@(C.Reg.AnyCFG cfg) = (cfgName cfg, c)
   in Map.fromList (List.map cfgWithName (CSyn.parsedProgCFGs prog))
+
+-----
+-- Parsing overrides YAML files
+-----
+
+-- | The parsed contents of an overrides-related @.yaml@ file. These contents
+-- have not yet been checked for validity. For the validated version, see
+-- 'ResolvedOverridesYaml'.
+newtype ParsedOverridesYaml = ParsedOverridesYaml
+  { getParsedOverridesYaml :: Map.Map Word64 W4.FunctionName }
+  deriving stock Show
+  deriving newtype (Semigroup, Monoid)
+
+-- | The contents of an overrides-related @.yaml@ file, which have been
+-- validity-checked to ensure that all of the addresses exist in the binary and
+-- that all of the override names actually exist. For the pre-validated version,
+-- see 'ParsedOverridesYaml'.
+newtype ResolvedOverridesYaml w = ResolvedOverridesYaml
+  { getResolvedOverridesYaml :: Map.Map (MM.MemSegmentOff w) W4.FunctionName }
+  deriving stock Show
+  deriving newtype (Semigroup, Monoid)
+
+-- | Parse an overrides-related @.yaml@ file.
+parseOverridesYaml ::
+  FilePath ->
+  IO ParsedOverridesYaml
+parseOverridesYaml yamlPath = do
+  res <- Yaml.decodeFileEither yamlPath
+  case res of
+    Left ex -> throw $ GreaseException $ Text.pack $ Yaml.prettyPrintParseException ex
+    Right val -> parseFunctionAddressOverrides val
+
+-- | Validate the contents of an overrides-related @.yaml@ file. See the
+-- Haddocks for 'ResolvedOverridesYaml' for a description of what validity
+-- checks are performed.
+resolveOverridesYaml ::
+  MM.MemWidth w =>
+  MML.LoadOptions ->
+  MM.Memory w ->
+  -- | The set of override names.
+  Set W4.FunctionName ->
+  ParsedOverridesYaml ->
+  IO (ResolvedOverridesYaml w)
+resolveOverridesYaml loadOpts mem fnOvNames (ParsedOverridesYaml parsedFunAddrs) = do
+  resolvedFunAddrs <-
+    traverse
+      (\(addr, funName) -> do
+        let addrWord = MM.memWord $ addr + loadOffset
+        addrSegOff <-
+          case Loader.resolveAbsoluteAddress mem addrWord of
+            Just addrSegOff -> pure addrSegOff
+            Nothing -> throw $ GreaseException $
+              "Could not resolve " <> tshow addrWord <>
+              "to an address in the binary."
+        unless (Set.member funName fnOvNames) $
+          throw $ GreaseException $
+            "Could not find an override for a function named '" <>
+            W4.functionName funName <> "'."
+        pure (addrSegOff, funName))
+      (Map.toList parsedFunAddrs)
+  pure $ ResolvedOverridesYaml $ Map.fromList resolvedFunAddrs
+  where
+    loadOffset = fromMaybe 0 $ MML.loadOffset loadOpts
+
+-- | Helper, not exported.
+--
+-- Parse the contents of an overrides-related @.yaml@ file from a 'Aeson.Value'.
+parseFunctionAddressOverrides ::
+  Aeson.Value ->
+  IO ParsedOverridesYaml
+parseFunctionAddressOverrides val = do
+  obj <- asObject val
+  case KeyMap.lookup "function address overrides" obj of
```

Comment:
Looks like this will silently do nothing if the user misspells the key - perhaps we should check for unexpected keys?

Reply:
I'm of two minds on this. On the one hand, I like the idea of having the ability to pass an empty `.yaml` file and have it produce an empty `Map`. (This would be especially nice if we add more features later to `overrides.yaml` and want the ability to opt in to certain features.) On the other hand, you are right that if the user misspells `function address overrides`, then this will silently drop the overrides.

As an alternative proposal, perhaps we should have another check that sees if there are any non-`function address overrides` keys in the top-level YAML object and throw an error if so. This would still allow the user to omit `function address overrides` if they wish while disallowing obvious typos.

Reply:
Sounds good to me!

Reply:
Potentially worth UX tests for the misspelled key and empty YAML cases.